pkg/sql/vecindex/high_speed_backfill_design.md

```diff
@@ -0,0 +1,1102 @@
+2. **Batch-Based Refinement**: Conducts subsequent scans that process batches of random vectors (typically 1000 per batch) to refine the initial centroids using mini-batch k-means until convergence
+3. **Memory Management**: Maintains configurable memory limits (default 600MB) with LRU caching and disk spilling for large datasets
+4. **Prefixed Index Support**: Handles prefixed indexes by maintaining separate centroid sets for each unique prefix
+5. **Convergence Detection**: Monitors centroid changes across iterations to determine when the clustering has stabilized
+
+This phase is the most resource-intensive part of the process, as it involves multiple scans over potentially large datasets and iterative k-means computation.
+
+#### STRUCTURE_BUILT Phase  
+
+This phase indicates that the partition structure has been successfully constructed and is ready for use:
+
+1. **Complete Hierarchy Available**: All leaf and interior partitions have been created with stable centroids
+2. **Interior Partition Construction**: Builds the hierarchical tree structure on top of the leaf centroids, creating interior partitions that enable efficient tree traversal
+3. **Temporary Index Configuration**: Sets up temporary indexes to use the same partition structure as the main index, avoiding metadata duplication
+4. **Fixup Avoidance**: Ensures background fixup processes do not interfere with the carefully constructed structure
+5. **Ready for Backfill**: The index is now ready to transition to `BACKFILL_ONLY` where standard bulk loading can begin
+
+The transition from `STRUCTURE_BUILT` to `BACKFILL_ONLY` marks the handoff from vector-specific construction logic to the standard, high-performance bulk loading infrastructure.
+
+## Integration with Existing Systems
+
+### Schema Changer Integration
+
+The new structure building phases integrate cleanly with the existing declarative schema changer framework:
+
+#### New Schema Operations
+Two new operation types are added to handle vector index structure building:
+- `BuildVectorIndexStructure`: Performs the actual k-means clustering and partition construction
+- `MarkVectorIndexStructureBuilt`: Marks the structure building as complete and ready for backfill
+
+#### Status Transitions
+The new statuses follow existing schema changer patterns, with dependency rules ensuring proper ordering and atomicity. The schema changer's existing transaction and rollback mechanisms handle failures during structure building.
+
+#### Vector-Specific Logic Isolation
+Only vector indexes follow the new `STRUCTURE_BUILDING` → `STRUCTURE_BUILT` path. Regular secondary indexes bypass these phases entirely, maintaining full backward compatibility.
+
+### C-SPANN K-Means Integration
+
+The structure building algorithm leverages existing, proven components from the `pkg/sql/vecindex/cspann` package:
+
+#### Reused Components
+- **BalancedKmeans**: Binary k-means implementation adapted for hierarchical partition construction
```

Comment:
Actually, I don't think we can reuse the `BalancedKMeans` component; the mini-batch K-means algorithm will be different enough that it's likely to be simpler to just do that separately (e.g. K can be > 2 in the mini-batch case, we probably want different memory management, etc).